src/server/saved_objects/service/lib/repository.ts

```diff
@@ -0,0 +1,531 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import uuid from 'uuid';
+
+import { EsMappings, getRootType } from '../../../mappings';
+import { decorateEsError } from './decorate_es_error';
+import * as errors from './errors';
+import { includedFields } from './included_fields';
+import { getSearchDsl } from './search_dsl';
+import { trimIdPrefix } from './trim_id_prefix';
+
+// BEWARE: The SavedObjectClient depends on the implementation details of the SavedObjectsRepository
+// so any breaking changes to this repository are considered breaking changes to the SavedObjectsClient.
+
+interface AnyObject {
+  [key: string]: any;
+}
+
+export interface SavedObjectsRepositoryOptions {
+  index: string;
+  mappings: EsMappings;
+  callCluster: <T extends AnyObject>(method: string, params?: AnyObject) => Promise<T>;
+  onBeforeWrite: () => void | Promise<void>;
+}
+
+interface UpdateOptions {
+  version?: number;
+}
+
+interface BulkGetObject {
+  id: string;
+  type: string;
+}
+
+interface BulkCreateObject {
+  type: string;
+  id: string;
+  attributes: AnyObject;
+}
+
+interface BulkCreateOptions {
+  overwrite?: boolean;
+}
+
+interface NotFoundResponse {
+  found: false;
+  status?: number;
+}
+
+interface FindOptions {
+  type?: string | string[];
+  search?: string;
+  /**
+   * see Elasticsearch Simple Query String "query" argument for more information
+   */
+  searchFields?: string[];
+  page?: number;
+  perPage?: number;
+  sortField?: string;
+  sortOrder?: string;
+  fields?: string[];
+}
+
+function isNotFoundResponse(response: any): response is NotFoundResponse {
+  const docNotFound = response.found === false;
+  const indexNotFound = response.status === 404;
+  return docNotFound || indexNotFound;
+}
+
+export class SavedObjectsRepository {
+  private index: string;
+  private mappings: EsMappings;
+  private type = getRootType(this.mappings);
+  private onBeforeWrite: SavedObjectsRepositoryOptions['onBeforeWrite'];
+  private unwrappedCallCluster: SavedObjectsRepositoryOptions['callCluster'];
+
+  constructor(options: SavedObjectsRepositoryOptions) {
+    const {
+      index,
+      mappings,
+      callCluster,
+      onBeforeWrite = () => {
+        // noop
+      },
+    } = options;
+
+    this.index = index;
+    this.mappings = mappings;
+    this.onBeforeWrite = onBeforeWrite;
+    this.unwrappedCallCluster = callCluster;
+  }
+
+  /**
+   * Persist an object
+   * @param type
+   * @param attributes
+   * @param options
+   */
+  public async create(
+    type: string,
+    attributes: AnyObject = {},
+    options: { id?: string; overwrite?: boolean } = {}
+  ) {
+    const { id, overwrite = false } = options;
+
+    const method = id && !overwrite ? 'create' : 'index';
+    const time = this.getCurrentTime();
+
+    try {
+      interface Response {
```

Comment:
I wanted to avoid leaking `any` as much as possible, so each of the methods in `SavedObjectsRepository` casts the response from `callCluster()` to some type which feeds the return type for these methods. Not a huge fan of hard coding the response type here, but it seems to be at least as good as what we had, and better because TypeScript understands it and verifies that we stick to it.